ntoskrnl/ke/i386/mp.c

```diff
@@ -0,0 +1,127 @@
+        ProcessorState.SpecialRegisters.Idtr.Limit = IdtDesc.Limit;
+        ProcessorState.SpecialRegisters.Idtr.Base = (ULONG)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS) + GdtDesc.Limit + 1);
+
+        /* Prep Cr Regsters */
+        ProcessorState.SpecialRegisters.Cr0 = __readcr0();
+        ProcessorState.SpecialRegisters.Cr2 = __readcr2();
+        ProcessorState.SpecialRegisters.Cr3 = __readcr3();
+        ProcessorState.SpecialRegisters.Cr4 = __readcr4();
+
+          /* Prepare Segment Registers */
+        ProcessorState.ContextFrame.SegCs = KGDT_R0_CODE;
+        ProcessorState.ContextFrame.SegSs = KGDT_R0_DATA;
+        ProcessorState.ContextFrame.SegDs = KGDT_R0_DATA;
+        ProcessorState.ContextFrame.SegEs = KGDT_R0_DATA;
+
+        /* Clear GS */
+        ProcessorState.ContextFrame.SegGs = 0;
+        /* Set FS to PCR */
+        ProcessorState.ContextFrame.SegFs = KGDT_R0_PCR;
+
+        /* Load TSR */
+        ProcessorState.SpecialRegisters.Tr = KGDT_TSS;
+
+        ProcessorState.ContextFrame.Eip = (ULONG_PTR)KiSystemStartup;
+
+        /* Prep GDT and IDT based on BSPs */
+        PKGDTENTRY pGdt = (PKGDTENTRY)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS));
+        PKIDTENTRY pIdt = (PKIDTENTRY)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS) + GdtDesc.Limit + 1);
+
+        /* Allocate PCR - TSS - InitalThread */
+        PKPCR pKPcr = (PKPCR)PAPInfo;
+        PKTSS pKTss = (PKTSS)((ULONG_PTR)PAPInfo + sizeof(KPCR));
+        PKTHREAD InitialThread = (PKTHREAD)((ULONG_PTR) PAPInfo + sizeof(KPCR) + sizeof(KTSS));
+
+        KernelStack = MmCreateKernelStack(FALSE, 0);
+        ASSERT(KernelStack);
+
+        DPCStack = MmCreateKernelStack(FALSE, 0);
+        if (!DPCStack)
+        {
+            MmDeleteKernelStack(KernelStack, FALSE);
+        }
+
+        /* Initalize a new PCR for the specific AP */
+        KiInitializePcr(ProcessorCount,
+                     (PKIPCR)pKPcr,
+                     pIdt, 
+                     pGdt, 
+                     pKTss,
+                     InitialThread,
+                     (PVOID)DPCStack);
+
+        /* Prep a new loaderblock for AP */
+        KeLoaderBlock->KernelStack = (ULONG_PTR)KernelStack;
+        KeLoaderBlock->Prcb = (ULONG_PTR)pKPcr;
+        KeLoaderBlock->Thread = (ULONG_PTR)InitialThread;
+
+        /* Fully initalize AP's TSS */
+        Ki386InitializeTss(pKTss, pIdt, pGdt);
+
+    } while (HalStartNextProcessor(KeLoaderBlock, &ProcessorState));
+
+    ProcessorCount--;
+    /* Started means the AP itself is online, this doesn't mean it's seen by the kernel */
+    DPRINT1("HalStartNextProcessor: Sucessful AP startup count is %X\n", ProcessorCount);
+    /* We have finished starting processors, Time to cleanup! */
+    MmFreeContiguousMemory(PAPInfo);
+    MmDeleteKernelStack(KernelStack, FALSE);
+    MmDeleteKernelStack(DPCStack, FALSE);
```

Comment:
This shouldn't be freed on success. Each CPU needs the stuff in there.

Reply:
This is ran an extra time for a CPU that doesn't exist, 
if HalStartNextProcessor returns false, we didn't successfully start an AP.
So what should we do with this memory that never got an AP assigned to it?

Reply:
Only in case of failure to start an AP, should that per-CPU memory be deleted.

Reply:
That's whats happening here is it not?

Reply:
It would have been more obvious if the code was looking like:
```c
while (TRUE /* or some hardcoded limitation on the number of CPUs */)
{
    // Initialize the per-CPU data...
    // Call KiInitializePcr() and the other helpers...

    if (!HalStartNextProcessor(KeLoaderBlock, &ProcessorState))
    {
        /* Cannot start anymore CPUs: cleanup and bail out */
        // Delete the two allocated stacks.
        // Delete PAPInfo
        break;
    }

    /* The CPU started: perform additional initialization */

    // Wait for the CPU to actually startup into KiSystemStartup
    while (/* some condition */)
    {
        // Do something....
    }
    if (/* CPU still not started */)
    {
        // Tell it somehow to stop whatever it is doing
        // Do some cleanup for this CPU, but continue the main loop.
    }
}
```